src/node.zig

```diff
@@ -195,39 +209,44 @@ fn decode_from_object(comptime T: type, env: c.napi_env, object: c.napi_value) !
     };
 }
 
-fn decode_from_array(comptime T: type, env: c.napi_env, array: c.napi_value) ![]u8 {
+pub fn decode_events(
+    env: c.napi_env,
+    array: c.napi_value,
+    operation: Operation,
+    output: []u8,
+) !usize {
+    return switch (operation) {
+        .create_accounts => try decode_events_from_array(env, array, Account, output),
+        .create_transfers => try decode_events_from_array(env, array, Transfer, output),
+        .commit_transfers => try decode_events_from_array(env, array, Commit, output),
+        .lookup_accounts => try decode_events_from_array(env, array, u128, output),
+        else => unreachable,
+    };
+}
+
+fn decode_events_from_array(
+    env: c.napi_env,
+    array: c.napi_value,
+    comptime T: type,
+    output: []u8,
+) !usize {
     const array_length = try translate.array_length(env, array);
     if (array_length < 1) return translate.throw(env, "Batch must contain at least one event.");
 
-    const allocator = std.heap.c_allocator;
-    const events = allocator.alloc(T, array_length) catch {
-        return translate.throw(env, "Failed to allocate memory for batch in client.");
-    };
-    errdefer allocator.free(events);
+    const body_length = @sizeOf(T) * array_length;
+    if (@sizeOf(Header) + body_length > config.message_size_max) {
+        return translate.throw(env, "Batch is larger than the maximum message size.");
+    }
+
+    var results = std.mem.bytesAsSlice(T, output);
```

Comment:
In Zig's std lib, `std.mem.bytesAsSlice` has this implementation:

```zig
/// Given a slice of bytes, returns a slice of the specified type
/// backed by those bytes, preserving pointer attributes.
pub fn bytesAsSlice(comptime T: type, bytes: anytype) BytesAsSliceReturnType(T, @TypeOf(bytes)) {
    // let's not give an undefined pointer to @ptrCast
    // it may be equal to zero and fail a null check
    if (bytes.len == 0) {
        return &[0]T{};
    }

    const cast_target = CopyPtrAttrs(@TypeOf(bytes), .Many, T);

    return @ptrCast(cast_target, bytes)[0..@divExact(bytes.len, @sizeOf(T))];
}
```

That final `@divExact()` requires that our `output` slice must be an exact multiple of `T`, so I think we need to slice on output before passing to `bytesAsSlice` to guarantee that, otherwise we may run into a safety-checked runtime panic for non-multiple sizes, even if the `body_length` does not overflow the buffer.